halo1/model.py

```diff
@@ -46,27 +51,126 @@ def read_halo1model(filepath):
 
 		return jms
 
-def import_halo1_nodes_from_jms(jms, *, scale=1.0, node_size=0.02):
+
+from mathutils import Vector
+
+def import_halo1_nodes_from_jms(jms, *,
+		scale=1.0,
+		node_size=0.02,
+		max_attachment_distance=0.00001,
+		attach_bones=('bip01',)):
 	'''
-	Import all the nodes from a jms into the scene and returns a dict of them.
+	Import all the nodes from a jms into the scene as an armature and returns
+	a dict of them.
+
+	node_size is currently unused and may be depricated.
+
+	max_attachment_distance is the max distance a bone may deviate from
+	the line that signifies the direction of the parent.
+
+	attach_bones is a tuple of prefixes that this function should attempt to
+	connect.
+
+	Returns the armature object and a dict of index - bone pairs.
 	'''
+	view_layer = bpy.context.view_layer
+
 	scene_nodes = dict()
+
+	absolute_transforms = get_absolute_node_transforms_from_jms(jms.nodes)
+
+	armature = bpy.data.armatures.new('imported')
+	armature_obj = bpy.data.objects.new('imported', armature)
+
+	# We need to specifically link the object to this so we can use mode_set
+	# to set Blender to edit mode.
+
+	view_layer.active_layer_collection.collection.objects.link(armature_obj)
+	view_layer.objects.active = armature_obj
+
+	# We need to be in edit mode in order to be able to edit armatures at all.
+
+	edit_bones = armature.edit_bones
+	bpy.ops.object.mode_set(mode='EDIT')
+
+	# "directions" of where each node is pointing for the attachment progress.
+	# These really are just points that are offset 1000 units in the direction
+	# the node is pointing from the point of the node's in scene head.
+	# These are used so we can check if two bones should be attached.
+	directions = {}
+
 	for i, node in enumerate(jms.nodes):
-		scene_node = create_sphere(
-			name=NODE_NAME_PREFIX+node.name, size=node_size)
+		scene_node = edit_bones.new(name=NODE_NAME_PREFIX+node.name)
 
 		# Assign parent if index is valid.
 		scene_node.parent = scene_nodes.get(node.parent_index, None)
 
-		set_rotation_from_jms(scene_node, node)
+		# If a node has multiple children we cannot connect its tail end
+		# to the children.
+
+		pos = absolute_transforms[i]['translation']
+		rot = absolute_transforms[i]['rotation']
+
+		# Bones when created start at 0 0 0 for their tail and head.
+		# We extend the tail so it sticks out a bit and signifies the direction.
+
+		scene_node.tail = (scene_node.bbone_x, 0.0, 0.0)
```

Comment:
We're just extending the bone in its length axis so its size isn't 0, 0 sized bones are deleted.

This also is so that it points into a certain direction. A direction we later update to be the proper direction using the transform.